cudax/test/containers/async_buffer/assign.cu

```diff
@@ -0,0 +1,266 @@
+    }
+
+    { // cudax::async_buffer::assign(count, const T&), shrinking to empty
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(0, T(42));
+      CUDAX_CHECK(buf.empty());
+      CUDAX_CHECK(buf.data() != nullptr);
+    }
+
+    { // cudax::async_buffer::assign(count, const T&), shrinking
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(2, T(42));
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_size_value(buf, 2, T(42)));
+    }
+
+    { // cudax::async_buffer::assign(count, const T&), growing
+      Buffer buf{env, 4, T(-2)};
+      buf.assign(6, T(42));
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_size_value(buf, 6, T{42}));
+    }
+  }
+
+  SECTION("cudax::async_buffer::assign(iter, iter) forward iterators")
+  {
+    const cuda::std::array<T, 6> input = {T(1), T(42), T(1337), T(0), T(12), T(-1)};
+    { // cudax::async_buffer::assign(iter, iter), with forward iterators empty range
+      Buffer buf{env};
+      buf.assign(input.begin(), input.begin());
+      CUDAX_CHECK(buf.empty());
+      CUDAX_CHECK(buf.data() == nullptr);
+    }
+
+    { // cudax::async_buffer::assign(iter, iter), with forward iterators shrinking to empty
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(input.begin(), input.begin());
+      CUDAX_CHECK(buf.empty());
+      CUDAX_CHECK(buf.data() != nullptr);
+    }
+
+    { // cudax::async_buffer::assign(iter, iter), with forward iterators shrinking
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(input.begin(), input.end());
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_range(buf));
+    }
+
+    { // cudax::async_buffer::assign(iter, iter), with forward iterators growing
+      Buffer buf{env, 4, T(-2)};
+      buf.assign(input.begin(), input.end());
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_range(buf));
+    }
+  }
+
+  SECTION("cudax::async_buffer::assign(initializer_list)")
+  {
+    { // cudax::async_buffer::assign(initializer_list), empty range
+      Buffer buf{env};
+      buf.assign(cuda::std::initializer_list<T>{});
+      CUDAX_CHECK(buf.empty());
+      CUDAX_CHECK(buf.data() == nullptr);
+    }
+
+    { // cudax::async_buffer::assign(initializer_list), shrinking to empty
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(cuda::std::initializer_list<T>{});
+      CUDAX_CHECK(buf.empty());
+      CUDAX_CHECK(buf.data() != nullptr);
+    }
+
+    { // cudax::async_buffer::assign(initializer_list), shrinking
+      Buffer buf{env, 10, T(-2)};
+      buf.assign(cuda::std::initializer_list<T>{T(1), T(42), T(1337), T(0), T(12), T(-1)});
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_range(buf));
+    }
+
+    { // cudax::async_buffer::assign(initializer_list), growing
+      Buffer buf{env, 4, T(-2)};
+      buf.assign(cuda::std::initializer_list<T>{T(1), T(42), T(1337), T(0), T(12), T(-1)});
+      CUDAX_CHECK(!buf.empty());
+      CUDAX_CHECK(equal_range(buf));
+    }
+  }
+
+#if 0 // Implement exceptions
```

Comment:
in `__async/sender/exception.cuh` i have `_CUDAX_TRY` and `_CUDAX_CATCH` macros that use `NV_IF_TARGET` under the covers to either `try`/`catch` or `if constexpr (true)`/`if constexpr (false)` for host and device respectively. maybe those macros would be helpful in `async_buffer`?

Reply:
I would push that as a followup